File: python/src/semantic_retrieval/generator/completion_models/completion_model.py
from typing import Generic, Optional

from semantic_retrieval.common.types import P, R, Record
from semantic_retrieval.prompts.prompt import IPrompt
from semantic_retrieval.utils.callbacks import CallbackManager


class CompletionModelParams(Generic[P], Record):
    def __init__(
        self,
        prompt: str | IPrompt,
        model: Optional[str] = None,
        completion_params: Optional[P] = None,
    ):
        self.prompt = prompt
        self.model = model
        self.completion_params = completion_params


class CompletionModel(Generic[P, R]):
    callback_manager = None

    def __init__(self, callback_manager: Optional[CallbackManager] = None):
        self.callback_manager = callback_manager

    async def run(self, params: CompletionModelParams[P]) -> R:
        raise NotImplementedError

File: typescript/src/generator/completion-models/completionModel.ts
import { IPrompt } from "../../prompts/prompt";
import { CallbackManager, Traceable } from "../../utils/callbacks";

export interface CompletionModelParams<P = unknown> {
  prompt: string | IPrompt;
  model?: string;
  completionParams?: P;
}

export interface CompletionModelResponse<R = unknown> {
  data: R;
}

/**
 * A simple class for interacting with different LLM completion models. CompletionModels
 * are leveraged by CompletionGenerators to generate completions from prompts.
 */
export abstract class CompletionModel implements Traceable {
  callbackManager?: CallbackManager;

  constructor(callbackManager?: CallbackManager) {
    this.callbackManager = callbackManager;
  }

  abstract run(params: CompletionModelParams): Promise<CompletionModelResponse>;
}
